git/operations.go

```diff
@@ -276,25 +275,60 @@ func changed(ctx context.Context, workingDir, ref string, subPaths []string) ([]
 	return splitList(out.String()), nil
 }
 
-func execGitCmd(ctx context.Context, args []string, config gitCmdConfig) error {
-	if trace {
-		print("TRACE: git")
-		for _, arg := range args {
-			print(` "`, arg, `"`)
+// Whilst debugging or developing, you may wish to filter certain git commands out of the logs.
+func exemptedTraceGitCommand(command string) bool {
+	blacklistedGitCommands := []string{
+		// To filter out a certain git subcommand add it here, e.g.:
+		// "config",
+	}
+	for _, blacklistedGitCommand := range blacklistedGitCommands {
+		if blacklistedGitCommand == command {
+			return true
 		}
-		println()
+	}
+	return false
+}
+
+func traceGitCommand(args []string, config gitCmdConfig, out string) string {
+	command := `git ` + strings.Join(args, " ")
+
+	output := strings.Trim(out, "\x00")
+	scanner := bufio.NewScanner(strings.NewReader(output))
+	lines := []string{}
+	for scanner.Scan() {
+		lines = append(lines, scanner.Text())
+	}
+	output = strings.Join(lines, `\n`)
+
+	// trim trailing newline often found in `git` responses if present
+	if len(output) > 0 && output[len(output)-1] == '\n' {
+		output = output[:len(output)-1]
 	}
 
+	return fmt.Sprintf(
+		"TRACE: command=%q output=%q dir=%q env=%q",
+		command,
+		output,
+		config.dir,
+		strings.Join(config.env, ","),
+	)
+}
+
+// execGitCmd runs a `git` command with the supplied arguments.
+func execGitCmd(ctx context.Context, args []string, config gitCmdConfig) error {
 	c := exec.CommandContext(ctx, "git", args...)
 
 	if config.dir != "" {
 		c.Dir = config.dir
 	}
 	c.Env = append(env(), config.env...)
-	c.Stdout = ioutil.Discard
+
+	traceOutput := &bytes.Buffer{}
+	c.Stdout = traceOutput
 	if config.out != nil {
-		c.Stdout = config.out
+		c.Stdout = io.MultiWriter(config.out, traceOutput)
```

Comment:
Shouldn't this also be conditional on `trace`?

Reply:
It might be clearer to set the writers up as before, then have a stanza conditional on `trace` that swaps them out.

Reply:
ok.  I avoided doing this because it's functionally equivalent to what was there before (`c.Stdout = ioutil.Discard`).  Writing to an writer that's never read from is the same as writing to ioutil.Discard.

I'll make this change now